diff --git a/src/internal/client/runtime.js b/src/internal/client/runtime.js
index f09aaf34eec5babc118de54ff9087b95eda69906..73e118c6310ac5384edbd1bf1b6ad5c447be4320 100644
--- a/src/internal/client/runtime.js
+++ b/src/internal/client/runtime.js
@@ -681,6 +681,11 @@ export function get(signal) {
 
 		var is_new = (derived.f & REACTION_RAN) === 0;
 
+		if (should_connect && !is_new) {
+			unfreeze_derived_effects(derived);
+			reconnect(derived);
+		}
+
 		if (is_dirty(derived)) {
 			if (should_connect) {
 				// set the flag before `update_derived`, so that the derived
@@ -691,10 +696,6 @@ export function get(signal) {
 			update_derived(derived);
 		}
 
-		if (should_connect && !is_new) {
-			unfreeze_derived_effects(derived);
-			reconnect(derived);
-		}
 	}
 
 	if (batch_values?.has(signal)) {
