diff --git a/src/internal/client/runtime.js b/src/internal/client/runtime.js
index 27def053008cc438af3186a5d939668c786e871d..0f4392265ea77f231eb8138a9bd86305552b2496 100644
--- a/src/internal/client/runtime.js
+++ b/src/internal/client/runtime.js
@@ -686,10 +686,15 @@ export function get(signal) {
 			(derived.f & CONNECTED) === 0 &&
 			!untracking &&
 			active_reaction !== null &&
-			(is_updating_effect || (active_reaction.f & CONNECTED) !== 0);
+			(active_reaction.f & CONNECTED) !== 0;
 
 		var is_new = (derived.f & REACTION_RAN) === 0;
 
+		if (should_connect && !is_new) {
+			unfreeze_derived_effects(derived);
+			reconnect(derived);
+		}
+
 		if (is_dirty(derived)) {
 			if (should_connect) {
 				// set the flag before `update_derived`, so that the derived
@@ -700,10 +705,6 @@ export function get(signal) {
 			update_derived(derived);
 		}
 
-		if (should_connect && !is_new) {
-			unfreeze_derived_effects(derived);
-			reconnect(derived);
-		}
 	}
 
 	if (batch_values?.has(signal)) {
